### DISCUSSION TITLE
-resize PNG's & preserve XMP metadata

### DISCUSSION THREAD

**Discussion by user-1138** [ImageMagick/ImageMagick#4114] [Help]
Hi ya'll!

I've been working on a WordPress/Imagick PHP Extension issue [over here](https://wordpress.stackexchange.com/questions/393437/how-to-control-wordpress-image-metadata-using-imagick), which led me today to test ImageMagick locally/directly and produced essentially the same results (no matter what all PNG metadata is stripped. And JPG metadata is preserved unless the -strip operator is set). 

Which I guess leads me to this fundamental question - 
**1. How can I resize PNG images whilst also preserving their metadata?**
And secondly - 
**2. How can I replicate this behavior in the PHP imagick extension?**

Thanks in advance!

**PNG metadata stripped**
```
convert meta-test.png -resize 100x100 new.png
convert meta-test.png -strip -resize 100x100 new.png
```
**JPG metadata preserved and then stripped**
```
convert meta-test.jpg -resize 100x100 new.jpg
convert meta-test.jpg -strip -resize 100x100 new.jpg
```

Here's the test JPG and PNG images I've been mucking with if you want something to test with.
[https://ibb.co/0jFBC56](https://ibb.co/0jFBC56)
[https://ibb.co/qRYb0g8](https://ibb.co/qRYb0g8)

**System:**
Debian GNU/Linux 11
ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25

**Comment by snibgo** [comment]
After using IM, you can use exiftool to copy metadata. For example:
```
magick meta-test.png -resize 100x100 new.png

exiftool -tagsFromFile meta-test.png new.png
```

**Comment by user-1138** [comment]
Thanks for the reply @snibgo!

I'm well aware of exiftool (I use it to set the PNG metadata in the first place), and yes I could FTP all the WordPress thumbnails down, set the metadata, and then ftp them back up, but... I'd prefer to avoid this tedious workaround if possible and instead get to the root of the issue.

So I want to know how/if possible to stop ImageMagick from stripping metadata out of PNG's, and if not possible/why not?

**Comment by snibgo** [comment]
As far as I know, we can't tell IM to copy all the metadata. I think the issue is that IM doesn't read all the metadata from PNG files. I don't know how difficult it would be to modify IM to do this.

**Comment by fmw42** [comment]
libpng does its own thing. I do not think ImageMagick can over-ride what libpng wants to do. Contact the libpng developers.

**Comment by user-1138** [comment]
@fmw42 @snibgo - I joined/asked the libpng **png-mng-implement** mailing list about this and their reply (verbatim) was: 

> This issue is related to ImageMagick and not libpng. The reason is that when a PNG is written, the metadata needs to be provided while writing it. Libpng is not stripping anything.
> 
> The possible cause could be inability of ImageMagick to write the metadata in PNG, or because the operation you supplied modified parameters which are redundant with image properties, but the capability does not exist within ImageMagick to update the metadata properly to reflect the changes. It is likely decided to not output metadata which is now incorrect or misleading.

What do ye ImageMagician's make of that? :-)

**Comment by snibgo** [comment]
I wrote:
> I think the issue is that IM doesn't _read_ all the metadata from PNG files.

libpng mailing list wrote:
> The reason is that when a PNG is _written_, the metadata needs to be provided while writing it.

The reason I think the problem is with _reading_ comes from playing with your two files. Reading the JPEG and writing a JPEG or PNG has no problem; the metadata is written. Reading the PNG and writing a PNG or JPEG is a problem; the metadata is not copied.

I know nothing about metadata interaction between IM and libpng, so I don't know where a fix might be found.

EDIT: Of course, there are other possibilities. Perhaps IM correctly reads the metadata from PNG, but then "forgets" it. Some time would be needed to read/edit/trace/debug/document the IM and libpng code to discover how to fix the problem. I don't have that time.

**Comment by dlemstra** [comment]
ImageMagick was unable to read the xmp profile from the `iTXt` chunk because it was stored in an unsupported way. Support for this was just added to ImageMagick 7 but this will not be backported to ImageMagick 6 so you will need to upgrade ImageMagick if you want this new feature.

```console
> identify -format "%[profiles]" meta-test.png
xmp
> magick meta-test.png new.png
> exiftool new.png | grep Creator
Creator                         : George Jetson
```

**Reply by snibgo** [reply]
Thanks, Dirk.
